File: algorithm_java/Day2/boj_9012.java
package Day2;

/*
괄호 문제
 */

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class boj_9012 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int T = Integer.parseInt(br.readLine());

        while (T > 0) {
            var stk = new Stack<Character>();
            String s = br.readLine();

            boolean flag = true;
            for (int i = 0; i < s.length(); i++) {
                char c = s.charAt(i);

                if (c == '(') stk.push(c);
                else {
                    if (!stk.empty()) stk.pop();
                    else {
                        flag = false;
                        break;
                    }
                }
            }

            if (stk.empty() && flag) sb.append("YES");
            else sb.append("NO");
            sb.append("\n");

            T--;
        }

        System.out.println(sb);
        br.close();
    }
}

File: algorithm_cpp/Project1/Algorithm_Session/boj_9012.cpp
#include <iostream>
#include <stack>
#include <string>

using namespace std;

/*
괄호 문자열 문제
- Stack을 이용하여 해결
*/

void boj_9012() {
	ios::sync_with_stdio(false);
	cin.tie(0);

	int T;
	cin >> T;

	while (T--) {
		string s;
		cin >> s;
		
		stack<char> stk;
		bool flag = true;

		for (char c : s) {
			if (c == '(')
				stk.push(c);
			else {
				if (!stk.empty())
					stk.pop();
				else {
					flag = false;
					break;
				}
			}
		}

		if (stk.empty() && flag) cout << "YES" << "\n";
		else cout << "NO" << "\n";
	}
}
